OpenRVS/classes/OpenHTTPClient.uc

```diff
@@ -0,0 +1,261 @@
+
+	if (CallbackName == "")
+		class'OpenLogger'.static.Debug("callbackName should not be empty", self);
+
+	parsed = parseHttpUrl(url);
+	Host = parsed.Host;
+	Port = parsed.Port;
+	Path = parsed.Path;
+	class'OpenLogger'.static.Debug("sending request to http://" $ Host $ ":" $ Port $ Path, self);
+
+	BindPort();//create client socket
+	class'OpenLogger'.static.Debug("client tcp socket created", self);
+	LinkMode = MODE_Text;
+	ReceiveMode = RMODE_Event;
+	class'OpenLogger'.static.Debug("starting dns resolution", self);
+	Resolve(parsed.Host);//also sends request
+}
+
+// Separates an HTTP URL into Host, Port, and Path.
+private static function ParsedURL parseHttpUrl(string url)
+{
+	local array<string> urlParts;
+	local int numParts;
+	local ParsedURL parsed;
+	local int i;
+	local string path;
+	local int sep;//index of ":"
+
+	if (InStr(url, "http://") == -1)
+		url = "http://" $ url;//make http:// optional
+
+	//before: "http://www.example.com[:80]/servers"
+	//after: ["http:", "", "www.example.com[:80]", "servers"]
+	numParts = class'OpenString'.static.Split(url, "/", urlParts);
+
+	// Parse host and port.
+	if (InStr(urlParts[2], ":") == -1)//url does not contain port
+	{
+		parsed.Host = urlParts[2];
+		parsed.Port = 80;//make port optional
+	}
+	else//url contains port
+	{
+		parsed.Host = Left(urlParts[2], InStr(urlParts[2], ":"));//up to :
+		parsed.Port = int(Mid(urlParts[2], InStr(urlParts[2], ":")+1));//after :
+	}
+
+	// Parse path.
+	for (i=3; i<numParts; i++)//i=3 skips proto and hostport, starting at path
+	{
+		path = path $ "/" $ urlParts[i];
+	}
+	if (path == "")
+		path = "/";//make path optional
+	parsed.Path = path;	
+
+	return parsed;
+}
+
+// Parses an HTTP response string, returning the code and message body.
+private static function HttpResponse parseHttpResponse(string response)
+{
+	local string CRLF;
+	local HttpResponse resp;
+	local int bodyStartsAt;
+
+	// HTTP response format:
+	// First line is "HTTP/1.1 <HTTP CODE>" + CRLF e.g. "HTTP/1.1 200 OK"
+	// Next N lines are info+headers followed by CRLF.
+	// Next line is CRLF only.
+	// Remaining lines are HTTP response body.
+
+	if (Left(response, 4) != "HTTP")//this is not an HTTP response
+	{
+		resp.Code = -1;
+		return resp;
+	}
+	resp.Code = int(Mid(response, 9, 11));//response code is always these three characters.
+
+	// Two CRLF in a row precede response body.
+	CRLF = Chr(13) $ Chr(10);
+	bodyStartsAt = InStr(response, CRLF $ CRLF);
+	resp.Body = Mid(response, bodyStartsAt+4);//4=bytes of CR LF CR LF
+
+	return resp;
+}
+
+//
+// Callbacks (in lieu of waiting for responses, get a callback instead)
+// Callbacks should accept an HttpResponse as the only param and return nothing.
+//
+
+// Feed data to the server list parser.
+ function ServerListCallback(HttpResponse resp)
+{
+	if (ServerListCallbackProvider == none)
+	{
+		class'OpenLogger'.static.Debug("ServerListCallbackProvider was none", self);
+		return;//nothing to do
```

Comment:
Requiring a ServerListCallbackProvider which is always an OpenServerList means that this OpenHTTPClient class is less useful for future work than it could be.  This means that we ALWAYS need to be creating this class from an OpenServerList, but OpenServerList has a specific use and what if we want to use OpenHTTPClient in a different context in the future?

What do you think of adding a way to expand functionality of this class.  Maybe a function that returns the received HTTP response as a string, so if we create an OpenHTTPClient for a different need in the future and don't have a ServerListCallbackProvider, we can call this function to get the received data anyways.

Reply:
Originally I was thinking the process would look like this:

1. Think of a new use case for OpenHTTPClient
2. Add a CallbackProvider var and Callback function to OpenHTTPClient.
3. Register the Callback in Closed().

So the HTTP client would grow with a list of different callbacks for different use cases. Since the HTTP client is event-driven and not request-response driven, the callbacks help us correlate the response to the original requester. My first version had a `Get()` function which returned the string, but I had difficulty pulling the data from `Closed()` into `Get()` without the callback.

When creating OpenHTTPClient from the new class, ServerListCallbackProvider can be `none` as it won't be used.

Reply:
Perhaps I should document the above at a minimum